text/0000-ques-in-main.md

```diff
@@ -112,80 +215,184 @@ signature:
-    fn write_diagnostic(&self, progname: &str, stream: &mut Write) -> ();
-    fn exit_status(&self) -> i32;
+    fn report(self) -> i32;
 }
 ```
 
-`write_diagnostic` shall write a complete diagnostic message for
-`self` to its `stream` argument.  If it produces no output, there will
-be no output.  `stream` will be the same stream that `panic` messages
-from the main thread would go to, which is normally "standard error".
-(An application-controllable override mechanism may make sense,
-[see below][squelching-diagnostics].)
-
-The `progname` argument is a short version of the program's name
-(abstractly, what you would type at the command line to start the
-program, assuming it were in `PATH`; concretely, the basename of
-`argv[0]`, with any trailing `.exe` or `.com` chopped off on Windows,
-but not on other platforms).  This makes it easy to generate
-diagnostics in the style conventional for "Unixy" systems, e.g.
-
-    grep: /etc/shadow: Permission denied
-
-`exit_status` shall convert `self` to a platform-specific exit code,
-conveying at least a notion of success or failure.  The return type is
-`i32` to match [std::process::exit][] (which probably calls the C
-library's `exit` primitive), but (as already documented for
-`process::exit`) on "most Unix-like" operating systems, only the low 8
-bits of this value are significant.
+`report` is a call-once function; it consumes self.  The runtime
+guarantees to call this function after `main` returns, but at a point
+where it is still safe to use `eprintln!` or `io::stderr()` to print
+error messages.  `report` is not _required_ to print error messages,
+and if it doesn't, nothing will be printed.  The value it returns will
+be passed to `std::process::exit`, and shall convey at least a notion
+of success or failure.  The return type is `i32` to match
+[std::process::exit][] (which probably calls the C library's `exit`
+primitive), but (as already documented for `process::exit`) on "most
+Unix-like" operating systems, only the low 8 bits of this value are
+significant.
 
 [std::process::exit]: https://doc.rust-lang.org/std/process/fn.exit.html
 
-## Changes to `main`
-[changes-to-main]: #changes-to-main
+## Standard impls of Termination
+[standard-impls-of-termination]: #standard-impls-of-termination
 
-From the perspective of the code that calls `main`, its signature is
-now generic:
+At least the following implementations of Termination are available in
+libstd.  I use the ISO C constants `EXIT_SUCCESS` and `EXIT_FAILURE`
+for exposition; they are not necesssarily intended to be the exact
+values passed to `process::exit`.
 
 ``` rust
-fn<T: Termination> main() -> T { ... }
+impl Termination for ! {
+    fn report(self) -> i32 { unreachable!(); }
+}
+
+impl Termination for () {
+    fn report(self) -> i32 { EXIT_SUCCESS }
+}
+
+impl Termination for std::process::ExitStatus {
+    fn report(self) -> i32 {
+        self.code().expect("Cannot use a signal ExitStatus to exit")
```

Comment:
`ExitStatus::code` returns `Option<i32>`. I have to do _something_ for None, and None should only happen for ExitStatus values representing exit-on-signal, so this seemed logical.  Do you have a better idea?